File: Daily/904.fruit-into-baskets.js
/*
 * @lc app=leetcode id=904 lang=javascript
 *
 * [904] Fruit Into Baskets
 */

// @lc code=start
/**
 * @param {number[]} fruits
 * @return {number}
 */
var totalFruit = function (fruits) {
  // find the longest subarray having <=2 types of fruits
  const N = fruits.length;

  let left = 0;
  let right = 0;

  let maxLen = 0;

  /** @type {Map<number, number>} */
  const fruitTypesToNumber = new Map();

  // slide window. maintain fruitTypesToNumber.size <= 2
  while (right < N) {
    fruitTypesToNumber.set(
      fruits[right],
      (fruitTypesToNumber.get(fruits[right]) ?? 0) + 1,
    );

    if (fruitTypesToNumber.size > 2) {
      maxLen = Math.max(maxLen, right - left);

      while (fruitTypesToNumber.size > 2) {
        // move left
        fruitTypesToNumber.set(
          fruits[left],
          (fruitTypesToNumber.get(fruits[left]) ?? 0) - 1,
        );

        if (fruitTypesToNumber.get(fruits[left]) === 0) {
          fruitTypesToNumber.delete(fruits[left]);
        }

        left++;
      }
    }

    right++;
  }

  maxLen = Math.max(maxLen, right - left);

  return maxLen;
};
// @lc code=end

totalFruit([1, 2, 3, 4, 5]);

File: src/2025/904.fruit-into-baskets.cpp
/*
 * @lc app=leetcode id=904 lang=cpp
 *
 * [904] Fruit Into Baskets
 */

#include <unordered_map>
#include <vector>

// @lc code=start
class Solution {
 public:
  int totalFruit(const std::vector<int>& fruits) {
    std::unordered_map<int, int> fruit_to_counts;
    int left = 0;
    int right = 0;
    int max_fruit_count = 0;
    int current_fruit_count = 0;

    while (right < fruits.size()) {
      if (fruit_to_counts.size() == 2 &&
          !fruit_to_counts.contains(fruits[right])) {
        max_fruit_count = std::max(max_fruit_count, current_fruit_count);

        while (fruit_to_counts.size() == 2) {
          fruit_to_counts[fruits[left]]--;
          current_fruit_count--;
          if (fruit_to_counts[fruits[left]] == 0) {
            fruit_to_counts.erase(fruits[left]);
          }
          left++;
        }
      }

      current_fruit_count++;
      fruit_to_counts[fruits[right]]++;
      right++;
    }

    max_fruit_count = std::max(max_fruit_count, current_fruit_count);

    return max_fruit_count;
  }
};
// @lc code=end
